### PR TITLE
refactor: merge optional and required parameters in to one type calle…

### PR DESCRIPTION
**Pull request by tonsV2**
The goal of this PR is to merge optional and required parameters into one type

The loader still have traces of the two old types but with the coming stacks rewrite it'll probably change drastically so I didn't bother updating it.

### CODE REVIEW

File: pkg/instance/repository.go [hunk 1]
```diff
@@ -229,58 +227,34 @@ func groupWithInstances(instancesMap map[string][]*model.Instance, groupMap map[
...
-			requiredParameters[i].StackName = instance.StackName
-		}
-	}
-
-	optionalParameters := instance.OptionalParameters
-	if len(optionalParameters) > 0 {
-		for i := range optionalParameters {
-			optionalParameters[i].InstanceID = instance.ID
-			optionalParameters[i].StackName = instance.StackName
+	Parameters := instance.Parameters
```
Comment (radnov) [inline pkg/instance/repository.go:230]: `parameters` 👀

File: pkg/instance/repository.go [hunk 2]
```diff
@@ -229,58 +227,34 @@ func groupWithInstances(instancesMap map[string][]*model.Instance, groupMap map[
 
 // TODO: Rename PopulateRelations? Or something else?
```
Comment (radnov) [inline pkg/instance/repository.go:228]: this PR seems like a good opportunity to get rid of the TODO - maybe `populateParameters`?

File: pkg/instance/service.go [hunk 3]
```diff
@@ -187,24 +167,7 @@ func (s service) unlink(id uint) error {
...
-		if _, ok := parameterNames[parameter.StackRequiredParameterID]; !ok {
-			unmatchedParameters = append(unmatchedParameters, parameter.StackRequiredParameterID)
-		}
-	}
-
-	return unmatchedParameters
-}
-
-func matchOptionalParameters(stackParameters []model.StackOptionalParameter, instanceParameters []model.InstanceOptionalParameter) []string {
+func matchOptionalParameters(stackParameters []model.StackParameter, instanceParameters []model.InstanceParameter) []string {
```
Comment (radnov) [inline pkg/instance/service.go:170]: should be `matchParameters`, right?

File: pkg/model/instance.go [hunk 4]
```diff
@@ -34,38 +33,20 @@ type Linked struct {
...
-	for _, parameter := range i.RequiredParameters {
-		if parameter.StackRequiredParameterID == name {
+func (i Instance) FindParameter(name string) (InstanceParameter, error) {
+	for _, parameter := range i.Parameters {
+		if parameter.StackParameterID == name {
 			return parameter, nil
 		}
 	}
-	return InstanceRequiredParameter{}, fmt.Errorf("required parameter not found: %s", name)
+	return InstanceParameter{}, fmt.Errorf("required parameter not found: %s", name)
```
Comment (radnov) [inline pkg/model/instance.go:42]: Did you intentionally leave "required" in the error here?

File: pkg/model/instance.go [hunk 5]
```diff
@@ -34,38 +33,20 @@ type Linked struct {
...
-
-type InstanceOptionalParameter struct {
+type InstanceParameter struct {
 	InstanceID uint `json:"-" gorm:"primaryKey"`
-	// TODO: Rename StackOptionalParameterID to Name
-	StackOptionalParameterID string                 `json:"name" gorm:"primaryKey"`
-	StackOptionalParameter   StackOptionalParameter `json:"-" gorm:"foreignKey:StackOptionalParameterID,StackName; references:Name,StackName; constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
-	StackName                string                 `json:"-"`
-	Value                    string                 `json:"value"`
+	// TODO: Rename StackParameterID to Name
```
Comment (radnov) [inline pkg/model/instance.go:47]: does it make sense to do this as well with this PR?

File: pkg/model/stack.go [hunk 6]
```diff
@@ -7,38 +7,32 @@ import (
...
-	HostnamePattern    string                   `json:"hostnamePattern"`
-	HostnameVariable   string                   `json:"hostnameVariable"`
+	Name             string           `json:"name" gorm:"primaryKey"`
+	Parameters       []StackParameter `json:"parameters" gorm:"foreignKey:StackName; references: Name; constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
+	Instances        []Instance       `json:"instances" gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
+	HostnamePattern  string           `json:"hostnamePattern"`
+	HostnameVariable string           `json:"hostnameVariable"`
+
+	CreatedAt time.Time `json:"createdAt"`
+	UpdatedAt time.Time `json:"updatedAt"`
```
Comment (radnov) [inline pkg/model/stack.go:17]: Moving these Gorm fields to the bottom kind of breaks the "convention" of having them at the top like we do for the other models. I'm okay either way, but just have it consistent.

File: pkg/model/stack.go [hunk 7]
```diff
@@ -7,38 +7,32 @@ import (
...
 		}
 	}
-	return StackOptionalParameter{}, fmt.Errorf("optional parameter not found: %s", name)
-}
-
-type StackRequiredParameter struct {
-	Name      string `json:"name" gorm:"primaryKey"`
-	StackName string `json:"stackName" gorm:"primaryKey"`
-	Consumed  bool   `json:"consumed"`
+	return StackParameter{}, fmt.Errorf("optional parameter not found: %s", name)
```
Comment (radnov) [inline pkg/model/stack.go:30]: no need for "optional" here as well, right?

File: scripts/instances/deploy-dhis2-db.sh [hunk 8]
```diff
@@ -26,7 +26,7 @@ echo "{
       \"value\": \"$DATABASE_ID\"
     }
   ],
-  \"optionalParameters\": [
+  \"parameters\": [
```
Comment (radnov) [inline scripts/instances/deploy-dhis2-db.sh:29]: should remove `requiredParameters` above

**Review by radnov — COMMENTED** [review]
Nice work, using just one type for the parameters makes a lot of sense 👍

File: pkg/model/instance.go [hunk 9]
```diff
@@ -7,23 +7,22 @@ import (
...
+	ID          uint                `json:"id" gorm:"primarykey"`
+	User        User                `json:"user"`
+	UserID      uint                `json:"userId"`
+	Name        string              `json:"name" gorm:"index:idx_name_and_group,unique"`
+	Group       Group               `json:"group"`
+	GroupName   string              `json:"groupName" gorm:"index:idx_name_and_group,unique"`
+	Description string              `json:"description"`
+	StackName   string              `json:"stackName"`
+	TTL         uint                `json:"ttl"`
+	Parameters  []InstanceParameter `json:"Parameters" gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```
Comment (teleivo) [inline pkg/model/instance.go:19]:
```suggestion
	Parameters  []InstanceParameter `json:"parameters" gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```

File: pkg/instance/helmfile.go [hunk 10]
```diff
@@ -130,13 +130,14 @@ func configureInstanceEnvironment(accessToken string, instance *model.Instance,
...
-	for _, parameter := range instance.OptionalParameters {
-		instanceEnv := fmt.Sprintf("%s=%s", parameter.StackOptionalParameter.Name, parameter.Value)
+	for _, parameter := range stack.Parameters {
+		instanceParameter, err := instance.FindParameter(parameter.Name)
+		if err == nil {
+			instanceEnv := fmt.Sprintf("%s=%s", instanceParameter.StackParameter.Name, instanceParameter.Value)
+			cmd.Env = append(cmd.Env, instanceEnv)
+			continue
+		}
+		instanceEnv := fmt.Sprintf("%s=%s", parameter.Name, *parameter.DefaultValue)
```
Comment (teleivo) [inline pkg/instance/helmfile.go:140]: What if there is no default value? If whatever comes next can never succeed this should turn into an error.

File: pkg/storage/postgresql.go [hunk 11]
```diff
@@ -20,7 +20,7 @@ func NewDatabase(c config.Postgresql) (*gorm.DB, error) {
 	dsn := fmt.Sprintf("host=%s user=%s password=%s dbname=%s port=%d sslmode=disable", host, username, password, name, port)
 
 	databaseConfig := gorm.Config{
-		Logger:         logger.Default.LogMode(logger.Warn),
+		Logger:         logger.Default.LogMode(logger.Info),
```
Comment (teleivo) [inline pkg/storage/postgresql.go:23]: Why did you change to `Info`? Not too noisy to have it on permanently?

File: pkg/model/instance.go (hunk 5 above)
Comment (tonsV2) [reply pkg/model/instance.go:47]: No, after merging the database service I didn't fix all the relations which obviously should be done but it's beyond the scope of this PR

File: pkg/stack/loader.go [hunk 12]
```diff
@@ -120,13 +114,16 @@ func parseStack(dir, name string) (*stack, error) {
 	requiredParams := extractRequiredParameters(file, stackParameters)
 	optionalParams := extractOptionalParameters(file, stackParameters)
 
+	for _, parameter := range requiredParams {
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply pkg/stack/loader.go:117]: I'm adding all the required parameters to the optional parameters and finally storing them as (just) parameters

File: pkg/instance/helmfile.go (hunk 10 above)
Comment (tonsV2) [reply pkg/instance/helmfile.go:140]: Why? Won't it just be nothing and that's ok?

**Review by teleivo — APPROVED** [review]

File: pkg/instance/service.go [hunk 13]
```diff
@@ -213,20 +176,18 @@ func matchOptionalParameters(stackParameters []model.StackOptionalParameter, ins
...
 }
 
 func validateParameters(stack *model.Stack, instance *model.Instance) error {
-	unmatchedRequiredParameters := matchRequiredParameters(stack.RequiredParameters, instance.RequiredParameters)
-	unmatchedOptionalParameters := matchOptionalParameters(stack.OptionalParameters, instance.OptionalParameters)
+	unmatchedOptionalParameters := matchParameters(stack.Parameters, instance.Parameters)
 
 	unmatchedParameters := make([]string, 0)
-	unmatchedParameters = append(unmatchedParameters, unmatchedRequiredParameters...)
 	unmatchedParameters = append(unmatchedParameters, unmatchedOptionalParameters...)
```
Comment (radnov) [inline pkg/instance/service.go:191]: Can we just replace that section with:
```
unmatchedParameters := matchParameters(stack.Parameters, instance.Parameters)
```

**Review by radnov — APPROVED** [review]
